Src/Framework/Units.py

```diff
@@ -0,0 +1,816 @@
+
+            u = copy.deepcopy(self)
+            u._div_by(self._out_factor)
+
+            s = fmt.format(u._factor)
+            s += ' {}'.format(self._ounit)
+
+        return s
+
+    def _div_by(self, other):
+
+        if self.is_analog(other):
+            self._factor /= other._factor
+            self._dimension = [0,0,0,0,0,0,0,0,0]        
+        elif self._equivalent:
+            equivalence_factor = self.get_equivalence_factor(other)
+            if equivalence_factor is not None:
+                self._factor /= other._factor/equivalence_factor
+                self._dimension = [0,0,0,0,0,0,0,0,0]
+            else:
+                raise UnitError('The units are not equivalent')
+        else:
+            self._factor /= other._factor
+            for i in range(len(self._dimension)):
+                self._dimension[i] = self._dimension[i] - other._dimension[i]
+
+        self._ounit = None
+        self._out_factor = None
+
+    def _mult_by(self, other):
+
+        if self.is_analog(other):
+            self._factor *= other._factor
+            for i in range(len(self._dimension)):
+                self._dimension[i] = 2.0*self._dimension[i]
+        elif self._equivalent:
+            equivalence_factor = self.get_equivalence_factor(other)
+            if equivalence_factor is not None:
+                self._factor *= other._factor/equivalence_factor
+                for i in range(len(self._dimension)):
+                    self._dimension[i] = 2*self._dimension[i]
+                return
+            else:
+                raise UnitError('The units are not equivalent')
+        else:
+            self._factor *= other._factor
+            for i in range(len(self._dimension)):
+                self._dimension[i] = self._dimension[i] + other._dimension[i]
+
+        self._ounit = None
+        self._out_factor = None
+
+    def ceil(self):
+        """Ceil of a _Unit value in canonical units. 
+
+        >>> print(measure(10.2, 'm/s').ceiling())
+        10.0000 m / s
+        >>> print(measure(3.6, 'm/s').ounit('km/h').ceiling())
+        10.0 km / h
+        >>> print(measure(50.3, 'km/h').ceiling())
+        50.0 km / h
+        """
+
+        r = copy.deepcopy(self)
+
+        if r._ounit is not None:
+            val = math.ceil(r.toval(r._ounit))
+            newu = _Unit('au',val)
+            newu *= _str_to_unit(r._ounit)
+            return newu.ounit(r._ounit)
+        else:
+            r._factor = math.ceil(r._factor)
+            return r
+
+    @property
+    def dimension(self):
+        """Getter for _dimension attribute. Returns a copy.
+        """
+
+        return copy.copy(self._dimension)
+
+    @property
+    def equivalent(self):
+        """Getter for _equivalent attribute.
+        """
+
+        return self._equivalent
+
+    @equivalent.setter
+    def equivalent(self, equivalent):
+        """Setter for _equivalent attribute.
+        """
+
+        self._equivalent = equivalent
```

Comment:
Is there any point to having a `@property` wrapper and a `setter` if the setter simply acts as a wrapper, without calling any function in between the assignment? Why not just have a `self.equivalent` attribute without the `@property` instead of `self._equivalent` attribute and wrappers? 

This also goes for the other wrappers below, `format` and `units`. On the other hand, `dimension` and `factor` are ok since they don't have setters, so it's obvious that they have to be coded the way they are.

Reply:
I set properties for `_equivalent` because I always declared my variable as _hidden_ when designing a class. In the case of `_equivalent`, I agree that it is a bit overdesign but it allows to be consistent with other variables such as `dimension`.